src/Store/FileExplorerStore.re

```diff
@@ -6,102 +6,154 @@
-    });
-  };
+let start = () => {
+  let (stream, _) = Isolinear.Stream.create();
 
+  // TODO: remove when effect has been exposed wherever that ought to be (VimStore?)
   let openFileByPathEffect = path => {
     Isolinear.Effect.createWithDispatch(
       ~name="explorer.openFileByPath", dispatch => {
       dispatch(Actions.OpenFileByPath(path, None, None))
     });
   };
 
-  let updater = (state: State.t, action: Actions.t) => {
+  let updater = (state: State.t, action: FileExplorer.action) => {
+    let setTree = tree => {
+      ...state,
+      fileExplorer: {
+        ...state.fileExplorer,
+        tree: Some(tree),
+      },
+    };
+
+    let replaceNode = (nodeId, node) =>
+      switch (state.fileExplorer.tree) {
+      | Some(tree) =>
+        setTree(FsTreeNode.update(nodeId, ~tree, ~updater=_ => node))
+      | None => state
+      };
+
     switch (action) {
-    | OpenExplorer(directory) => (
-        state,
-        getExplorerFilesEffect(
-          directory,
-          state.languageInfo,
-          state.iconTheme,
-          Configuration.getValue(c => c.filesExclude, state.configuration),
-        ),
+    | TreeUpdated(tree) => (setTree(tree), Isolinear.Effect.none)
+
+    | NodeUpdated(id, node) => (
+        replaceNode(id, node),
+        Isolinear.Effect.none,
       )
 
-    | ExplorerNodeClicked(node) =>
+    | NodeClicked(node) =>
       switch (node) {
       | {kind: File, path, _} => (state, openFileByPathEffect(path))
 
       | {kind: Directory({children: `Loading}), _} => (
           state,
-          switch (state.fileExplorer.directory) {
-          | Some(tree) =>
-            updateFolderEffect(
-              ~folder=node.path,
-              ~nodeId=node.id,
-              ~languageInfo=state.languageInfo,
-              ~iconTheme=state.iconTheme,
-              ~ignored=
-                Configuration.getValue(
-                  c => c.filesExclude,
-                  state.configuration,
-                ),
-              ~tree,
-            )
-          | None => Isolinear.Effect.none
-          },
+          Effects.refreshNode(
+            node,
+            state.languageInfo,
+            state.iconTheme,
+            state.configuration,
+          ),
         )
 
-      | _ =>
-        switch (state.fileExplorer.directory) {
-        | Some(tree) =>
-          let tree = FsTreeNode.toggleOpenState(node.id, ~tree);
-          (state, setExplorerTreeEffect(tree));
-        | None => (state, Isolinear.Effect.none)
-        }
+      | _ => (
+          replaceNode(node.id, FsTreeNode.toggleOpenState(node)),
+          Isolinear.Effect.none,
+        )
       }
-
-    | _ => (state, Isolinear.Effect.none)
     };
   };
 
-  (updater, stream);
-};
+  (
+    (state: State.t) =>
+      fun
+      // TODO: Should be handle by a more general init mechanism
+      | Actions.Init => {
+          let cwd = Rench.Environment.getWorkingDirectory();
+          let newState = {
```

Comment:
This nested update always feels awkward - I wonder if lenses could streamline this? Like [`lens`](https://opam.ocaml.org/packages/lens/).

Reply:
I'm not sure lenses really solve the actual problem. Having deeply nested updates is usually an indication of architectural flaws, and using lenses is a bit like just brushing it under the carpet.

This isn't really a nested update though, just a value being set on a record that also happens to contain a record. We might make it a bit smoother by adding a `setWorkspace(cwd, state)` or `Workspace.create(cwd)` function for example.